### PR TITLE
proxy: allow disable protocol detection on specific ports

### LINKED ISSUE #339: MySQL and other server-speak-protocols require explicit configuration
Our protocol detection logic currently only works for client-speaks-first protocols--where the client immediately sends data upon the TCP connection completing. MySQL and some other protocols are server-speaks-first; the client won't send anything on the connection until the server sends something. We should change the protocol detection logic so that it is insensitive to which side speaks first, insofar as that is possible.

Currently in Conduit 0.2 we document that the user must explicitly configure the proxy to turn off L7 features for server-speaks-first protocols. This is counter to our goal to "just work" with zero configuration.

A potential strategy:

* Assume at first that the underlying service being proxied always speaks one protocol.
* Assume the underlying service speaks TCP (not HTTP) until we get a positive indication otherwise.
* When we haven't marked the underlying service as speaking HTTP, as soon as we receive an incoming connection or making an outgoing connection, start the connection to the destination (in the inbound case, this would be the connection to the underlying service; in the outbound case, this would be the connection to the external service). Note that this is insensitive to which side speaks first. Mark the protocol for this as "TBD".
* If/when a client speaks first on a "TBD" connection, run our HTTP (and eventually TLS) detection logic. If/when that logic indicates HTTP and/or gRPC then reconfigure the connection as a HTTP (and/or gRPC) connection and mark the underlying service as HTTP. In particular, add it to the pool of connections for L7 HTTP load balancing.

/cc @seanmonstar @olix0r @wmorgan

### PR DESCRIPTION
**Pull request by seanmonstar**
- Adds environment variables to configure a set of ports that, when an
  incoming connection has an SO_ORIGINAL_DST with a port matching, will
  disable protocol detection for that connection and immediately start a
  TCP proxy.
- Adds a default list of well known ports: SMTP and MySQL.

Closes #339

### CODE REVIEW

**Comment by olix0r** [comment]
@seanmonstar Correct, ports are less likely to be randomly assigned in kubernetes. This isn't the case in all environments, but it's a reasonable assumption for now.

File: proxy/src/lib.rs [hunk 1]
```diff
@@ -186,6 +187,14 @@ where
             "serving Prometheus metrics on {:?}",
             metrics_listener.local_addr(),
         );
+        info!(
+            "private listener disables protocol detection on ports {:?}",
```
Comment (briansmith) [inline proxy/src/lib.rs]: I suggest "Protocol detection disabled for outgoing destination ports {:?}":

1. "private listener" is not relevant to users.
2. "outgoing" indicates that it's for outgoing traffic.
3. "destination" indicates it's looking at the destination port, not the source port.

File: proxy/src/lib.rs [hunk 2]
```diff
@@ -186,6 +187,14 @@ where
...
+            config.private_ports_disable_protocol_detection,
+        );
+        info!(
+            "public listener disables protocol detection on ports {:?}",
```
Comment (briansmith) [inline proxy/src/lib.rs]: Ditto: "Protocol detection disabled for incoming destination ports {:?}":

File: proxy/src/lib.rs [hunk 3]
```diff
@@ -234,14 +244,15 @@ where
             let outgoing = Outbound::new(
                 bind,
                 control,
-                config.default_destination_namespace().to_owned(),
+                config.pod_namespace.to_owned(),
```
Comment (briansmith) [inline proxy/src/lib.rs]: This change shouldn't be part of this PR since it is out of scope.

File: proxy/src/lib.rs [hunk 4]
```diff
@@ -254,6 +265,7 @@ where
 
         let (_tx, controller_shutdown_signal) = futures::sync::oneshot::channel::<()>();
         {
+            let report_timeout = config.report_timeout;
```
Comment (briansmith) [inline proxy/src/lib.rs]: This change shouldn't be part of this PR since it is out of scope.

File: proxy/src/lib.rs [hunk 5]
```diff
@@ -282,7 +294,7 @@ where
                         telemetry,
                         control_host_and_port,
                         dns_config,
-                        config.report_timeout,
+                        report_timeout,
```
Comment (briansmith) [inline proxy/src/lib.rs]: This change shouldn't be part of this PR since it is out of scope.

File: proxy/src/config.rs [hunk 6]
```diff
@@ -135,6 +141,8 @@ pub const ENV_METRICS_LISTENER: &str = "CONDUIT_PROXY_METRICS_LISTENER";
 const ENV_PRIVATE_CONNECT_TIMEOUT: &str = "CONDUIT_PROXY_PRIVATE_CONNECT_TIMEOUT";
 const ENV_PUBLIC_CONNECT_TIMEOUT: &str = "CONDUIT_PROXY_PUBLIC_CONNECT_TIMEOUT";
 pub const ENV_BIND_TIMEOUT: &str = "CONDUIT_PROXY_BIND_TIMEOUT";
+pub const ENV_PRIVATE_PORTS_DISABLE_PROTOCOL_DETECTION: &str = "CONDUIT_PROXY_PRIVATE_PORTS_DISABLE_PROTOCOL_DETECTION";
+pub const ENV_PUBLIC_PORTS_DISABLE_PROTOCOL_DETECTION: &str = "CONDUIT_PROXY_PUBLIC_PORTS_DISABLE_PROTOCOL_DETECTION";
```
Comment (briansmith) [inline proxy/src/config.rs]: I think these should be s/PRIVATE/OUTGOING/ and s/PUBLIC/INCOMING/.

"private" and "public" are not user-facing concepts they are part of the internals of Conduit. When we write the documentation for this feature we will (presumably) document one flag as controlling protocol detection for "incoming traffic destined for the given ports" and "outgoing traffic destined for the given ports" and we should use that terminology for these settings.

File: proxy/src/transparency/server.rs [hunk 7]
```diff
@@ -91,6 +95,31 @@ where
         let local_addr = connection.local_addr().unwrap_or(self.listen_addr);
         let proxy_ctx = self.proxy_ctx.clone();
 
+        let disable_protocol_detection = orig_dst
+            .map(|addr| {
+                self.disable_protocol_detection_ports.contains(&addr.port())
+            })
+            .unwrap_or(false);
+
+        if disable_protocol_detection {
```
Comment (briansmith) [inline proxy/src/transparency/server.rs]: This duplicates the code starting at line 169.  Please refactor this so the code duplication is removed. That way we don't need to keep them in sync.

File: proxy/src/lib.rs (hunk 3 above)
Comment (seanmonstar) [reply proxy/src/lib.rs]: It was actually needed, because `IndexSet` is not `Copy`, and since one of is moved out of the `Config` earlier, it's no longer possible to call a method on `&Config`. Same with the `let report_timeout = ..` changes below.

File: proxy/src/config.rs (hunk 6 above)
Comment (seanmonstar) [reply proxy/src/config.rs]: Oh really? Have I misunderstood? I thought "private" config options configured our `Outbound` listener, and "public" configured our `Inbound` listener.

Comment (seanmonstar) [reply proxy/src/config.rs]: My understanding of the two terms is the opposite: PUBLIC refers to the listener that is listening publicly, that anything else can connect to, in order to talk to the app (so, our `Inbound` stack), and PRIVATE refers to the listener that only the app can talk to (so, our `Outbound` stack).

Comment (briansmith) [reply proxy/src/config.rs]: Maybe we should talk face-to-face about it since I think we have some miscommunication.

To backtrack a bit, you said earlier:
> I thought "private" config options configured our Outbound listener, and "public" configured our Inbound listener

However, that's not exactly the case. The `Inbound` stack is configured using `ENV_PUBLIC_LISTENER`, `ENV_PRIVATE_FORWARD`, and `ENV_PRIVATE_CONNECT_TIMEOUT`:

```rust
        let inbound_listener = BoundPort::new(config.public_listener.addr)
            .expect("public listener bind");

[...]

        let inbound = {
            [...]
            let default_addr = config.private_forward.map(|a| a.into());
            let fut = serve(
                config.private_connect_timeout,
                Inbound::new(default_addr, bind),
                [...])
            [...]
        };
```

Similarly, the `Outbound` stack is configured using `ENV_PRIVATE_LISTENER` and `ENV_PUBLIC_CONNECT_TIMEOUT`.

One way of thinking about it is that the `Inbound` stack is a pair (public listener, private destination) and the `Outbound` stack is a pair (private listener, public destinations). The new settings you are adding configure the behavior for the "public listener" and the "public destinations"; i.e. the "public" end of both stacks. That's why I said "private" vs. "public" is the wrong distinction: they're both "public". The distinction that matters is that one setting is for the public end of the `Inbound` stack and the other setting is for the public end of the `Outbound` stack.

Comment (briansmith) [reply proxy/src/config.rs]: I thought of a new way of explaining the naming rationale: `conduit inject` has these arguments:

```
--skip-inbound-ports uintSlice    ports that should skip the proxy and send directly to the application (default [])
--skip-outbound-ports uintSlice   outbound ports that should skip the proxy (default [])
```

proxy-init has these arguments:

```
--inbound-ports-to-ignore intSlice  
--incoming-proxy-port int             Port to redirect incoming traffic (default -1)
      --outbound-ports-to-ignore intSlice   Outbound ports to ignore and not redirect to proxy. This has higher precedence than any other parameters.
--outgoing-proxy-port
```

So "inbound" and "outbound" are the terms we're already using for the vary similar feature, and those are the terms that we'll almost definitely use in the `conduit inject` arguments that we'll add later. (I think it's odd that proxy-init uses both "inbound/outbound" & "incoming/outgoing".)

I think this is a strong point in favor of using the "inbound" and "outbound" terminology now.

File: proxy/src/transparency/server.rs (hunk 7 above)
Comment (seanmonstar) [reply proxy/src/transparency/server.rs]: I've refactored into a tiny function below.

**Comment by seanmonstar** [comment]
Alright, updated private/public usage to inbound/outbound.

File: proxy/tests/support/proxy.rs [hunk 8]
```diff
@@ -61,6 +65,16 @@ impl Proxy {
         self
     }
 
+    pub fn disable_inbound_ports_protocol_detection(mut self, ports: Vec<u16>) -> Self {
+        self.inbound_disable_ports_protocol_detection = Some(ports);
+        self
+    }
```
Comment (briansmith) [inline proxy/tests/support/proxy.rs]: Sorry i didn't notice this before. I think it would be better to just have the tests set `ENV_INBOUND_PORTS_DISABLE_PROTOCOL_DETECTION` directly, instead of offering this convenience API for them. That would more closely match what happens in the production code.

File: proxy/src/transparency/server.rs (hunk 7 above)
Comment (briansmith) [reply proxy/src/transparency/server.rs]: Thanks. I see that didn't work out quite as well as I expected it would but I guess it will still help with future changes.
